internal/pipeline/types.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// This file defines the type structures used by pipeline
+package pipeline
+
+import (
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Function defines an item in the pipeline function list
+type Function struct {
+	// Image is the image name that a function will use
+	Image string `yaml:"image,omitempty"`
+	// Config specifies an inline k8s resource used as the function config.
+	// Config, ConfigPath, and ConfigMap fields are mutually exclusive.
+	Config yaml.Node `yaml:"config,omitempty"`
+	// ConfigPath specifies a relative path to a file in the current directory
+	// containing a K8S resource used as the function config. This resource is
+	// excluded when resolving 'sources', and as a result cannot be operated on
+	// by the pipeline.
+	ConfigPath string `yaml:"configPath,omitempty"`
+	// ConfigMap is a convenient way to specify a function config of kind ConfigMap.
+	ConfigMap map[string]string `yaml:"configMap,omitempty"`
+}
+
+// Pipeline contains the content in pipeline.yaml file
```

Comment:
Please carry over the comments in go/kpt-pipeline-declaration, they're the authoritative description of the semantics of this resource.